packages/loader/container-loader/src/test/connectionStateHandler.spec.ts

```diff
@@ -1139,6 +1143,48 @@ describe("ConnectionStateHandler Tests", () => {
 		},
 	);
 
+	it("test 'read' reconnect & races ", async () => {
+		connectionStateHandler = createHandler(
+			false, // connectedRaisedWhenCaughtUp,
+			true, // readClientsWaitForJoinSignal
+		);
+
+		connectionStateHandler.establishingConnection({ text: "initial connect" });
+		connectionStateHandler.receivedConnectEvent(connectionDetails);
+		connectionStateHandler_receivedJoinSignalEvent(connectionDetails);
+		assert.strictEqual(
+			connectionStateHandler.connectionState,
+			ConnectionState.Connected,
+			"Client 1 should be in connected state",
+		);
+
+		connectionStateHandler.receivedDisconnectEvent({ text: "disconnect" });
+
+		connectionDetails2.mode = "read";
+		connectionStateHandler.establishingConnection({ text: "initial connect" });
+		connectionStateHandler.receivedConnectEvent(connectionDetails2);
+
+		// Clear Audience
+		connectionStateHandler_receivedLeaveSignalEvent(connectionDetails.clientId);
+
+		// We connected to different front-end, it still things that first client is there!
+		connectionStateHandler_receivedJoinSignalEvent(connectionDetails);
+		connectionStateHandler_receivedJoinSignalEvent(connectionDetails2);
+
+		// It should not wait for leave of connectionDetails.clientId
```

Comment:
I'm not sure... how do we know we can ignore that duplicate Join?  We have no indication it was duplicate.

In this test, the Leave signal triggers this code, right?

`this.applyForConnectedState("removeMemberEvent");`

So then the concept is that we have cleared that old client, and when the duplicate Join comes in, even though its clientId matches ours, we ignore it, because who cares.

Hmmm maybe it's ok.  Weird.

Reply:
It's not "duplicate" from Audience POV - Audience is cleared before we process "initial signals" on a new connection. But yes, same ID "joins" audience again.

There is nothing to ignore here by ConnectionStateHandler - it simple is indifferent to that event.
It's not really "ours", I think that's the key. ConnectionStateHandler should make right calls on loss of connection (RE wait for leave or not), but after that it operates only with pendingClientId.

That said, yes, I agree that applyForConnectedState() overall looks a bit wrong (in terms of else branch).
It's not just applyForConnectedState("removeMemberEvent"), even as unrelated as applyForConnectedState("containerSaved") seems like can trigger "connectedStateRejected" event that looks wrong to me.

Reply:
Ah, containerSaved() cancels timeout before calling applyForConnectedState("containerSaved").
Same for receivedRemoveMemberEvent()
And we produce error event only for "timeout", so that looks Ok (i.e. "connectedStateRejected" event is information only, and can show up number of times for same connection).